arngll/src/frame_info.rs

```diff
@@ -0,0 +1,690 @@
+    KeyIndex = 1,
+    Reserved2 = 2,
+    Reserved3 = 3,
+}
+
+impl KeyIdentMode {
+    pub fn try_from_u8(x: u8) -> Option<KeyIdentMode> {
+        match x {
+            0 => Some(Self::Addresses),
+            1 => Some(Self::KeyIndex),
+            2 => Some(Self::Reserved2),
+            3 => Some(Self::Reserved3),
+            _ => None,
+        }
+    }
+
+    pub fn to_u8(&self) -> u8 {
+        match self {
+            Self::Addresses => 0,
+            Self::KeyIndex => 1,
+            Self::Reserved2 => 2,
+            Self::Reserved3 => 3,
+        }
+    }
+}
+
+impl TryFrom<u8> for KeyIdentMode {
+    type Error = anyhow::Error;
+
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        KeyIdentMode::try_from_u8(value).ok_or(format_err!("{} is not a valid key ident mode", value))
+    }
+}
+
+impl From<KeyIdentMode> for u8 {
+    fn from(value: KeyIdentMode) -> u8 {
+        value.to_u8()
+    }
+}
+
+#[derive(Clone, Eq, PartialEq)]
+pub struct SecInfo {
+    pub enc: bool,
+    pub kim: KeyIdentMode,
+    pub fcntr: u32,
+    pub kid: Option<u8>,
+    pub mic: Mic,
+}
+
+impl SecInfo {
+    pub fn from_iter<'a, T: Iterator<Item=&'a u8>>(iter: &mut T) -> SecInfo {
+        let scf = iter.next().copied().unwrap();
+        let enc = (scf & 0b10000000) != 0;
+        let miclen = MicLen::try_from_u8((scf & 0b01100000) >> 5).unwrap();
+        let kim = KeyIdentMode::try_from_u8((scf & 0b00011000) >> 3).unwrap();
+        let fcntr = u32::from_be_bytes([
+            iter.next().copied().unwrap(),
+            iter.next().copied().unwrap(),
+            iter.next().copied().unwrap(),
+            iter.next().copied().unwrap()
+        ]);
+        let kid = if kim == KeyIdentMode::KeyIndex {
+            Some(iter.next().copied().unwrap())
+        } else {
+            None
+        };
+
+        SecInfo {
+            enc,
+            kim,
+            fcntr,
+            kid,
+            mic: Mic { len: miclen, .. Mic::EMPTY }
+        }
+    }
+
+    pub fn scf(&self) -> u8 {
+        (u8::from(self.enc) << 7)
+            + (self.mic.len.to_u8() << 5)
+            + (self.kim.to_u8() << 3)
+    }
+
+    pub fn bytes(&self) -> impl Iterator<Item=u8> {
+        once(self.scf())
+            .chain(self.fcntr.to_be_bytes())
+            .chain(self.kid)
+    }
+}
+
+impl Debug for SecInfo {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{{")?;
+        if self.enc {
+            write!(f, "ENC ")?;
+        }
+
+        write!(f, "FCNTR={} KIM={:?}",self.fcntr, self.kim)?;
+
+        if let Some(kid) = self.kid {
+            write!(f, " KID=0x{:02X}",kid)?;
+        }
+
+        write!(f, " MIC={:?}",self.mic)?;
+        write!(f, "}}")
+    }
+}
+
+
+#[derive(Clone, Eq, PartialEq)]
+pub struct FrameInfo {
+    pub frame_type: FrameType,
+    pub ack_requested: bool,
+    pub is_from_relay: bool,
+    pub network_id: Option<NetworkId>,
+    pub dst_addr: HamAddr,
```

Comment:
If these can be empty, as in the code below, why not make them `Option<HamAddr>` and replace that with `HamAddr::EMPTY` when consumed?